### DISCUSSION TITLE
How to use the "always" attribute?

### DISCUSSION THREAD

**Discussion by jcmcclorey** [rebing/graphql-laravel#866] [Q&A]
Laravel => 7
PHP => 7.2.34
rebing/graphql-laravel => 8.0

I'm looking for clarification on what the "always" attribute is for on Types. From reading the docs, I am led to believe that the fields in "always" are queried from the data source, but what I have tried hasn't worked so I'm not sure if I'm on the right track.

I'm attempting to use it as shown in the docs, with the only difference being that the fields in "always" are NOT fields defined in the GraphQL schema on the related Type.

I did also try to use "always" for fields defined on a related GraphQL Type, but didn't notice any difference.

**Comment by mfn** [comment]
> I'm looking for clarification on what the "always" attribute is for on Types.

I'll give it a try 🤞🏼 : 

When you decide to use `SelectFields` (i.e. you add a `'model'` attribute to your GraphQL type), the queries generated are based on the fields on your type(s) and only the ones used in the _Graphql query_ are actually requested in the _SQL query_, which means the it uses Eloquent to request your model from the database _but_ it only requests the fields = attributes specified.

But in GraphQL, you can have resolvers of `fieldA` which, internally, depend on `fieldB` **always** being present. Which means:
- the GraphQL query requests `fieldA`
- so via SQL and thus on the Eloquent model, only `fieldA` is loaded
- but the resolver of `fieldA` actually requires `fieldB` because there's some internal logic

This is now the use case to add `'always'` to `fieldB`.

We've also some tests covering this, e.g.
- `\Rebing\GraphQL\Tests\Database\SelectFields\AlwaysTests\AlwaysTest::testAlwaysSingleField`
  which uses the GraphQL type form
- `\Rebing\GraphQL\Tests\Database\SelectFields\AlwaysTests\PostType::fields`



> I'm attempting to use it as shown in the docs, with the only difference being that the fields in "always" are NOT fields defined in the GraphQL schema on the related Type.
> 
> I did also try to use "always" for fields defined on a related GraphQL Type, but didn't notice any difference.

I don't fully understand that part, what you tried. Specifically what means "related Type" in this context.

Please check the tests, maybe it helps?

**Reply by jcmcclorey** [reply]
@mfn Thanks for the effort. I definitely understand it more than I did before.

To answer your question about "related Type", using the example in the docs with a user and posts; (under [this section](https://github.com/rebing/graphql-laravel#eager-loading-relationships)) posts would be the "related Type". So in that example I am referring to the `fields()` defined on the `PostType`.

I also have one more follow-up question: Is there any way to specify certain database fields to always include in an Eloquent query? Not saying this needs to be added, just wondering if it already exists.

**Reply by mfn** [reply]
> Is there any way to specify certain database fields to always include in an Eloquent query?

You mean _outside_ of GraphQL fields and using them together with `'always'`?

I can think of two things:
- when you call `$getSelectFields()`, you receive a query object => you can mutate it before "get"ting the data
- It _might_ also be possible with `'query'` on a field: the closure you provide receives a Query builder and you can mutate it to your liking (I think you could even replace it); but that would require that field to be fetched _or_ used together with `'always'`, though not sure this works

---

I've tried neither so far:
- for one, I don't use SelectFields due to its hard to understand options/code/applicability
- further, when I work with Eloquent, I always completely load the models (not the relations, that indeed depends on the GraphQL tree being queried)
  Partially loaded Eloquent models pose such a risk of causing bugs that I entirely forgo using this for GraphQL. E.g. your model _getter_ could require another loaded column you forgot about and this isn't obvious when mapping your GraphQL fields to model attributes, etc.

Since I don't use SelectFields, I'm not really sure regarding the relation part (loading it without being requested). Maybe `'always'` works, maybe not. I exclusively use a [dataloader approach](https://www.google.com/search?hl=en&q=graphql%20dataloader%20php) (but whipped up my own approach) use the underlying `Deferred` of graphql-php to eager load requested relations. It's definitely less convenient than using SelectFields but gives me full control and insights into what's going on. 🤷🏼

**Reply by jcmcclorey** [reply]
OK, thanks for the explanations. You've been a big help.

I will see what I can come up with based on your suggestions.
